src/deriveNativeScriptHash.h

```diff
@@ -0,0 +1,51 @@
+#ifndef H_CARDANO_APP_DERIVE_NATIVE_SCRIPT_HASH
+#define H_CARDANO_APP_DERIVE_NATIVE_SCRIPT_HASH
+
+#include "bip44.h"
+#include "cardano.h"
+#include "common.h"
+#include "handlers.h"
+#include "nativeScriptHashBuilder.h"
+
+handler_fn_t deriveNativeScriptHash_handleAPDU;
+
+// a special type for distinguishing what to show in the UI, makes it easier
+// to handle PUBKEY DEVICE_OWNED vs THIRD_PARTY
+typedef enum {
+	UI_SCRIPT_PUBKEY_PATH = 0,
+	UI_SCRIPT_PUBKEY_HASH,
+	UI_SCRIPT_ALL,
+	UI_SCRIPT_ANY,
+	UI_SCRIPT_N_OF_K,
+	UI_SCRIPT_INVALID_BEFORE,
+	UI_SCRIPT_INVALID_HEREAFTER,
+} ui_native_script_type;
+
+typedef struct {
+	uint32_t totalScripts;
+	uint32_t remainingScripts;
+} complex_native_script_t;
+
+typedef union {
+	uint32_t requiredScripts;
+	bip44_path_t pubkeyPath;
+	uint8_t pubkeyHash[ADDRESS_KEY_HASH_LENGTH];
+	uint64_t timelock;
+} native_script_content_t;
+
+typedef struct {
+	uint8_t level;
+	// stores information about a complex script at the index level
+	complex_native_script_t complexScripts[MAX_SCRIPT_DEPTH];
+
+	uint8_t scriptHashBuffer[SCRIPT_HASH_LENGTH];
+	native_script_hash_builder_t hashBuilder;
+
+	native_script_content_t scriptContent;
+
+	// UI information
+	int ui_step;
+	ui_native_script_type ui_scriptType;
+} ins_derive_native_script_hash_context_t;
```

Comment:
Not that it's a problem necessary, but padding rules make this an inefficient layout.

Reply:
Hm, what would be a more efficient layout, or more precisely what's making this layout inefficient?

Reply:
https://developers.ledger.com/docs/nano-app/memory-alignment/
It'll pad out single bytes before larger (like here, 32 bit) types. If we fit into the memory that's fine, I doubt this is the largest struct in the union.

Reply:
I don't think we are concerned with memory here. (We run pretty close to limits for stuff with several layers of buffers, e.g. when we are deep within tx signing, and need to derive keys and encode addresses into bech32.)